flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -150,6 +160,59 @@ long getWindowTime() {
 		}
 
 		/**
+		 * Check pattern after match skip strategy.
+		 */
+		private void checkPatternSkipStrategy() {
+			AfterMatchSkipStrategy afterMatchSkipStrategy = currentPattern.getAfterMatchSkipStrategy();
+			if (afterMatchSkipStrategy.getStrategy() == AfterMatchSkipStrategy.SkipStrategy.SKIP_TO_FIRST ||
+				afterMatchSkipStrategy.getStrategy() == AfterMatchSkipStrategy.SkipStrategy.SKIP_TO_LAST) {
+				Pattern<T, ?> pattern = currentPattern;
+				while (!pattern.getName().equals(afterMatchSkipStrategy.getPatternName())) {
+					pattern = pattern.getPrevious();
+				}
+				// pattern name match check.
+				if (pattern == null) {
+					throw new MalformedPatternException("the pattern name specified in AfterMatchSkipStrategy " +
+						"can not be found in the given Pattern");
+				} else {
+					// can not be used with optional states.
+					if (pattern.getQuantifier().hasProperty(Quantifier.QuantifierProperty.OPTIONAL)) {
+						throw new MalformedPatternException("the AfterMatchSkipStrategy "
+							+ afterMatchSkipStrategy.getStrategy() + " can not be used with optional pattern");
+					}
+				}
+
+				// start position check.
+				if (pattern.getPrevious() == null) {
```

Comment:
Had a second thought on all those cases, and I think you had a good point with one of the previous proposals, just to fall back to the `SKIP_TO_NEXT_EVENT`

Maybe let's allow all those situations, but if within a match we cannot discard(e.g. skip is to the first pattern or the optional event is not present) then we discard nothing. The advantage in my opinion it is easier to understand than to understand and maintain all those cases. (As a side note the exceptions in SQL in all those cases in my opinion are just an implementation details as they keep a single partial match at a time). What do you think?